src/Compilers/Core/Portable/InternalUtilities/SetWithInsertionOrder.cs

```diff
@@ -0,0 +1,57 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Roslyn.Utilities
+{
+    /// <summary>
+    /// A set that returns the inserted values in insertion order.
+    /// The mutation operations are not thread-safe.
+    /// </summary>
+    internal sealed class SetWithInsertionOrder<T> : IEnumerable<T>
+    {
+        private Dictionary<T, uint> _map = new Dictionary<T, uint>();
```

Comment:
My intent was that the dictionary's values would be used later to implement `Remove`, but that doesn't really work (once you remove something all the indices shift). So I'll switch to a `HashSet`.